### DISCUSSION TITLE
QST: Serving geodataframe via matplotlib crashes flask server

### DISCUSSION THREAD

**Discussion by deeplook** [geopandas/geopandas#2732] [Q&A]
- [x] I have searched the [geopandas] tag on [StackOverflow](https://stackoverflow.com/questions/tagged/geopandas) and [GIS StackExchange](https://gis.stackexchange.com/questions/tagged/geopandas) for similar questions.

- [x] I have asked my usage related question on [StackOverflow](https://stackoverflow.com) or [GIS StackExhange](https://gis.stackexchange.com).

---

#### Question about geopandas

I need to return a bitmap image generated by geopandas from a geodataframe from a flask server. I managed to return a simple matplotlib plots, but when using geopandas I get: `... python3.10/site-packages/geopandas/plotting.py:673: UserWarning: Starting a Matplotlib GUI outside of the main thread will likely fail.` I've tried various things to avoid pyplot, but with no success. I've put the code into [this gist](https://gist.github.com/deeplook/f8208fbfbba94bf460a8903e978701f7). The crashing bit is below. Any help is greatly appreciated!

```python
@app.route('/geoplot')
def gdf_plot_png():
    # crashes the process
    path = geopandas.datasets.get_path("nybb")
    df = geopandas.read_file(path)
    df_wm = df.to_crs(epsg=3857)  # Web Mercator
    # The following line crashes with:
    #   ... python3.10/site-packages/geopandas/plotting.py:673: UserWarning: 
    #   Starting a Matplotlib GUI outside of the main thread will likely fail.
    ax = df_wm.plot(figsize=(4, 4), alpha=0.5, edgecolor="k")
    fig = ax.figure
    
    output = io.BytesIO()
    FigureCanvasAgg(fig).print_png(output)
    return Response(output.getvalue(), mimetype="image/png")
```

**Comment by jorisvandenbossche** [comment]
In what way does it crash? (the message about starting a matplotlib GUI is only a warning?)

I am not familiar with flask, but one thing might be to instruct matplotlib to not use a graphical backend before calling the geopandas plot method (so before importing matplotlib.pyplot). 

xref https://github.com/matplotlib/matplotlib/issues/12628

**Comment by deeplook** [comment]
Indeed, if I use `matplotlib.use("pdf")`, say, then flask will do the job, else it seems like another process is spawned, resulting in the (shortened) traceback below. And then, using fastapi insteaad with its `async` labels, there is no need for providing an alternative matplotlib backend.

```
[...]: UserWarning: Starting a Matplotlib GUI outside of the main thread will likely fail.
  fig, ax = plt.subplots(figsize=(10, 10))
2022-10-18 13:23:03.622 python[77558:2886321] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow drag regions should only be invalidated on the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00000001996c9148 __exceptionPreprocess + 240
	1   libobjc.A.dylib                     0x0000000199413e04 objc_exception_throw + 60
	2   CoreFoundation                      0x00000001996f40cc _CFBundleGetValueForInfoKey + 0
	3   AppKit                              0x000000019c1d7930 -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 372
	4   AppKit                              0x000000019c1c292c -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 948
	5   AppKit                              0x000000019c1c256c -[NSWindow initWithContentRect:styleMask:backing:defer:] + 56
	6   _macosx.cpython-38-darwin.so        0x0000000139c046f4 -[Window initWithContentRect:styleMask:backing:defer:withManager:] + 68
	7   _macosx.cpython-38-darwin.so        0x0000000139c07c70 FigureManager_init + 240
	8   python3.8                           0x0000000104b5ab04 wrap_init + 20
	9   python3.8                           0x0000000104aee9a0 wrapperdescr_call + 496
	10  python3.8                           0x0000000104ae21f4 _PyObject_MakeTpCall + 616
[...]
	131 python3.8                           0x0000000104ae2690 PyVectorcall_Call + 120
	132 python3.8                           0x0000000104cc7ac4 t_bootstrap + 196
	133 python3.8                           0x0000000104c5a414 pythread_wrapper + 48
	134 libsystem_pthread.dylib             0x000000019957c26c _pthread_start + 148
	135 libsystem_pthread.dylib             0x000000019957708c thread_start + 8
)
libc++abi: terminating with uncaught exception of type NSException
Command '['flask', 'run', '--port', '5001']' returned non-zero exit status 250.
```

**Comment by m-richards** [comment]
Running your gist, I'm not able to replicate the crash you describe (with or without setting the matplotlib backend). However I'm running windows, rather than macos.
